Sources/Datadog/DatadogConfiguration.swift

```diff
@@ -31,44 +31,126 @@ extension Datadog {
                 self.logsEndpoint = logsEndpoint
                 return self
             }
 
+            /// Sets the default service name associated with data send to Datadog.
+            /// NOTE: The `serviceName` can be also overwriten by each `Logger` instance.
+            /// - Parameter serviceName: the service name (default value is set to application bundle identifier)
+            public func set(serviceName: String) -> Builder {
+                self.serviceName = serviceName
+                return self
+            }
+
             /// Builds `Datadog.Configuration` object.
             public func build() -> Configuration {
                 return Configuration(
                     clientToken: clientToken,
-                    logsEndpoint: logsEndpoint
+                    logsEndpoint: logsEndpoint,
+                    serviceName: serviceName,
+                    environment: environment
                 )
             }
         }
     }
+
+    /// Valid SDK configuration, passed to the features.
+    ///
+    /// It takes two types received from the user: `Datadog.Configuration` and `AppContext` and blends them together
+    /// with resolving defaults and ensuring the configuration consistency.
+    internal struct ValidConfiguration {
+        internal let applicationName: String
+        internal let applicationVersion: String
+        internal let applicationBundleIdentifier: String
+        internal let serviceName: String
+        internal let environment: String
+
+        internal let logsUploadURLWithClientToken: URL
+
+        init(configuration: Datadog.Configuration, appContext: AppContext) throws {
+            self.init(
+                applicationName: appContext.bundleName ?? appContext.bundleType.rawValue,
+                applicationVersion: appContext.bundleVersion ?? "0.0.0",
+                applicationBundleIdentifier: appContext.bundleIdentifier ?? "unknown",
+                serviceName: configuration.serviceName ?? appContext.bundleIdentifier ?? "ios",
+                environment: try ifValid(environment: configuration.environment),
+                logsUploadURLWithClientToken: try ifValid(
+                    endpointURL: configuration.logsEndpoint.url,
+                    clientToken: configuration.clientToken
+                )
+            )
+        }
+
+        init(
+            applicationName: String,
+            applicationVersion: String,
+            applicationBundleIdentifier: String,
+            serviceName: String,
+            environment: String,
+            logsUploadURLWithClientToken: URL
+        ) {
+            self.applicationName = applicationName
+            self.applicationVersion = applicationVersion
+            self.applicationBundleIdentifier = applicationBundleIdentifier
+            self.serviceName = serviceName
+            self.environment = environment
+            self.logsUploadURLWithClientToken = logsUploadURLWithClientToken
+        }
```

Comment:
if we did
```swift
extension ValidConfiguration {
  init(configuration: Datadog.Configuration, appContext: AppContext) { ... }
}
```
we wouldn't need to write this `init(applicationName: ...) { ... }` as it would be generated by the compiler

Reply:
Done ✅. Although I don't like using extensions this way - it forces to put some part of the `struct` somewhere else in the file, and doesn't play well for readability in some cases (i.e. nested structs). Here it was relatively fine, as `ValidConfiguration` is the last type in this class extension.

Reply:
Definitely 💯 but this use case is measurably (metric: amount of less code) good without any downside, as long as we keep extension right below struct declaration

IMO this is one of the few (maybe the only one) good use cases of extensions